src/OpenLoco/Paint/PaintMiscEntity.cpp

```diff
@@ -0,0 +1,332 @@
+#include "PaintMiscEntity.h"
+#include "../CompanyManager.h"
+#include "../Graphics/Gfx.h"
+#include "../Graphics/ImageIds.h"
+#include "../Interop/Interop.hpp"
+#include "../Localisation/StringIds.h"
+#include "../Map/Tile.h"
+#include "Paint.h"
+#include <cassert>
+
+using namespace OpenLoco::Interop;
+
+namespace OpenLoco::Paint
+{
+    static void paintExhaustEntity(PaintSession& session, Exhaust* exhaustObject)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        if (dpi->zoom_level > 1)
+        {
+            return;
+        }
+        SteamObject* steamObject = exhaustObject->object();
+
+        uint8_t* edi = (exhaustObject->object_id & 0x80) == 0 ? steamObject->var_16 : steamObject->var_1A;
+        uint32_t imageId = edi[2 * exhaustObject->var_26];
+        imageId = imageId + steamObject->baseImageId + steamObject->var_0A;
+
+        if ((steamObject->var_08 & 8) != 0)
+        {
+            session.addToPlotListAsParent(imageId, { 0, 0, exhaustObject->z }, { 1, 1, 0 });
+        }
+        else
+        {
+            session.addToPlotListAsParent(imageId, { 0, 0, exhaustObject->z }, { 1, 1, 0 }, { -12, -12, exhaustObject->z });
+        }
+    }
+
+    static void paintRedGreenCurrencyEntity(PaintSession& session, MoneyEffect* moneyEffect)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        int8_t* wiggleYOffsets = reinterpret_cast<int8_t*>(0x4FAAC8);
+        if (dpi->zoom_level > 1)
+        {
+            return;
+        }
+        const string_id stringId = moneyEffect->amount >= 0 ? StringIds::format_currency_income_green : StringIds::format_currency_expense_red_negative;
+        uint32_t currencyAmount = abs(moneyEffect->amount);
+        int8_t* yOffsets = &wiggleYOffsets[moneyEffect->wiggle];
+
+        session.addToStringPlotList(currencyAmount, stringId, moneyEffect->y, moneyEffect->z, yOffsets, moneyEffect->offsetX);
+    }
+
+    static void paintWindowCurrencyEntity(PaintSession& session, MoneyEffect* moneyEffect)
+    {
+        Gfx::drawpixelinfo_t* dpi = session.getContext();
+        int8_t* wiggleYOffsets = reinterpret_cast<int8_t*>(0x4FAAC8);
```

Comment:
We should integrate these offsets as an `std::array`.

Reply:
That doesn't work. The array is 22 elements long which is then repeating few times and would need to make an implementation of draw function as well.

Reply:
This is the result with only 22 elements, doesn't work correctly:
![image](https://user-images.githubusercontent.com/8458425/107627489-475a8000-6c57-11eb-82d2-6a1c89ec0e06.png)

Reply:
You're right, it is much longer than I thought! I was tripped off by the negative elements. Marking them as signed leaves us with the following array:

```
0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3,
-3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2,
1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1,
2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3,
-3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0,
-1, -2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2,
3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3,
-2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1,
-2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3,
3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2,
-1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2,
-3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3,
2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3,
-3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2,
1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1,
2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3,
-3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0,
-1, -2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2,
3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3,
-2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1,
-2, -2, -3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3,
3, 3, 2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2,
-1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2,
-3, -3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3,
2, 2, 1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
0, 1, 2, 2, 3, 3, 3, 3, 2, 2, 1, 0, -1, -2, -2, -3,
-3, -3, -3, -2, -2, -1, 0, 1, 2, 2, 3, 3, 3, 3, 2, 2,
1, 0, -1, -2, -2, -3, -3, -3, -3, -2, -2, -1,
```

You're right that this repeats quite a few times, though. Let me see where it starts to cycle. We should be able to just apply a modulo to the index, right?

Reply:
Not currently, but after implementing sub functions yes (as it was in openrct2).

Reply:
Implemented now in the latest commit.